File: packages/examples/frameworks/vercel-ai-sdk/m3_multi_line.ts
/**
 * Example: Vercel AI SDK + Hymoex M3 (MoE MultiLine).
 *
 * Multiple teams using Vercel AI SDK's multi-step tool execution.
 * Each team exposes its experts as tools. The integrator selects
 * a team tool, which internally delegates to expert tools.
 *
 * Key difference from M2: M2 has flat tools. M3 uses team-level
 * tools that encapsulate their own expert tool sets — nested delegation.
 *
 * Requires: npm install ai @ai-sdk/openai zod
 */

import { generateText, tool } from "ai";
import { openai } from "@ai-sdk/openai";
import { z } from "zod";

// --- Hymoex M3: Manager -> Integrators -> Expert Manager -> Teams ---

// Finance team experts (internal to the team)
async function financeContracts(request: string): Promise<string> {
  const result = await generateText({
    model: openai("gpt-4o-mini"),
    system: "You are a contracts expert. Draft or review contracts concisely.",
    prompt: request,
  });
  return result.text;
}

async function financePayments(request: string): Promise<string> {
  const result = await generateText({
    model: openai("gpt-4o-mini"),
    system: "You are a payments expert. Process payroll and invoices concisely.",
    prompt: request,
  });
  return result.text;
}

// HR team experts (internal to the team)
async function hrOnboarding(request: string): Promise<string> {
  const result = await generateText({
    model: openai("gpt-4o-mini"),
    system: "You are an onboarding specialist. Handle new hire processes.",
    prompt: request,
  });
  return result.text;
}

async function hrCompliance(request: string): Promise<string> {
  const result = await generateText({
    model: openai("gpt-4o-mini"),
    system: "You are a compliance officer. Check regulatory requirements.",
    prompt: request,
  });
  return result.text;
}

// Team-level tools — each team encapsulates its own expert selection
// This is the M3 differentiator: tools that contain sub-teams

const financeTeamTool = tool({
  description:
    "Delegate to the finance team for contracts, payments, payroll, and invoicing",
  parameters: z.object({
    request: z.string(),
    subdomain: z.enum(["contracts", "payments"]),
  }),
  execute: async ({ request, subdomain }) => {
    // Team supervisor routes to the right expert within the team
    if (subdomain === "contracts") {
      return financeContracts(request);
    }
    return financePayments(request);
  },
});

const hrTeamTool = tool({
  description:
    "Delegate to the HR team for onboarding, compliance, and new hires",
  parameters: z.object({
    request: z.string(),
    subdomain: z.enum(["onboarding", "compliance"]),
  }),
  execute: async ({ request, subdomain }) => {
    if (subdomain === "onboarding") {
      return hrOnboarding(request);
    }
    return hrCompliance(request);
  },
});

// Integrator — selects which team handles the request
// Uses maxSteps for multi-step tool execution (team selection + expert execution)
async function integratorRoute(query: string) {
  const result = await generateText({
    model: openai("gpt-4o-mini"),
    system:
      "You are an integrator for enterprise onboarding. Route to the appropriate team and subdomain.",
    prompt: query,
    tools: {
      finance_team: financeTeamTool,
      hr_team: hrTeamTool,
    },
    maxSteps: 3, // Allow multi-step: team selection -> expert execution
  });
  return result;
}

async function main() {
  console.log("Vercel AI SDK + Hymoex M3: Nested tool delegation with maxSteps");
  console.log("Team tools encapsulate expert tools — true M3 composition");
  console.log("M2 = flat tools; M3 = team tools containing sub-experts");
  // const result = await integratorRoute("Draft the employment contract for John");
  // console.log(result);
}

main();

File: packages/examples/frameworks/pydantic-ai/m3_multi_line.py
"""Example: Pydantic AI + Hymoex M3 (MoE MultiLine).

Multiple teams with an integrator routing across them.
Each team has a supervisor agent that coordinates domain experts.

Requires: pip install pydantic-ai
"""

from pydantic_ai import Agent

import sys
sys.path.insert(0, "../../../../packages/hymoex-python/src")
from hymoex import (
    ExpertManagerSpec, ExpertSpec, IntegratorSpec, ManagerSpec,
    MultiLine, SupervisorSpec, Team, validate_topology,
)


# --- Hymoex M3: Manager -> Integrators -> Expert Manager -> Teams ---

topology = MultiLine(
    manager=ManagerSpec(objective="Enterprise onboarding"),
    integrators=[IntegratorSpec()],
    expert_manager=ExpertManagerSpec(),
    teams=[
        Team(name="finance", supervisor=SupervisorSpec(), experts=[
            ExpertSpec(domain="contracts"),
            ExpertSpec(domain="payments"),
        ]),
        Team(name="hr", supervisor=SupervisorSpec(), experts=[
            ExpertSpec(domain="onboarding"),
            ExpertSpec(domain="compliance"),
        ]),
    ],
)
print(f"Topology: M3, valid: {validate_topology(topology)['valid']}, agents: {topology.agent_count}")


# --- Pydantic AI Implementation ---

# Finance team experts
contracts_agent = Agent("openai:gpt-4o-mini", system_prompt="You are a contracts expert.")
payments_agent = Agent("openai:gpt-4o-mini", system_prompt="You are a payments expert.")

# HR team experts
onboarding_agent = Agent("openai:gpt-4o-mini", system_prompt="You are an onboarding specialist.")
compliance_agent = Agent("openai:gpt-4o-mini", system_prompt="You are a compliance officer.")

# Team supervisors
finance_supervisor = Agent("openai:gpt-4o-mini", system_prompt="You are a finance supervisor. Route to: contracts or payments.")
hr_supervisor = Agent("openai:gpt-4o-mini", system_prompt="You are an HR supervisor. Route to: onboarding or compliance.")

# Integrator — routes to the right team
integrator = Agent("openai:gpt-4o-mini", system_prompt="You are an integrator. Route to: finance or hr. Respond with one word.")


async def multi_line_route(query: str) -> str:
    """M3 pattern: Integrator -> Team Supervisor -> Expert."""
    # Integrator routes to team
    team_decision = await integrator.run(query)
    team = team_decision.data.strip().lower()

    # Team supervisor routes to expert within team
    if team == "finance":
        routing = await finance_supervisor.run(query)
        domain = routing.data.strip().lower()
        expert = contracts_agent if "contract" in domain else payments_agent
    else:
        routing = await hr_supervisor.run(query)
        domain = routing.data.strip().lower()
        expert = onboarding_agent if "onboard" in domain else compliance_agent

    result = await expert.run(query)
    return result.data


if __name__ == "__main__":
    import asyncio
    print("\nPydantic AI + Hymoex M3: Integrator -> Team Supervisors -> Experts")
    # Uncomment to run (requires OPENAI_API_KEY):
    # print(asyncio.run(multi_line_route("Set up payroll for new hire")))
